// @antv/l7-layers/lib/core/CommonStyleAttribute.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.getCommonStyleAttributeOptions = getCommonStyleAttributeOptions;
var _l7Core = require("@antv/l7-core");
function getCommonStyleAttributeOptions(name) {
  switch (name) {
    case 'stroke':
      return {
        name: 'stroke',
        type: _l7Core.AttributeType.Attribute,
        descriptor: {
          name: 'a_Stroke',
          buffer: {
            // give the WebGL driver a hint that this buffer may change
            usage: _l7Core.gl.DYNAMIC_DRAW,
            data: [],
            type: _l7Core.gl.FLOAT
          },
          size: 4,
          update: function update(feature) {
            var _feature$stroke = feature.stroke,
              stroke = _feature$stroke === void 0 ? [1, 1, 1, 1] : _feature$stroke;
            return stroke;
          }
        }
      };
    case 'opacity':
      return {
        name: 'opacity',
        type: _l7Core.AttributeType.Attribute,
        descriptor: {
          name: 'a_Opacity',
          buffer: {
            // give the WebGL driver a hint that this buffer may change
            usage: _l7Core.gl.STATIC_DRAW,
            data: [],
            type: _l7Core.gl.FLOAT
          },
          size: 1,
          update: function update(feature) {
            var _feature$opacity = feature.opacity,
              op = _feature$opacity === void 0 ? 1 : _feature$opacity;
            return [op];
          }
        }
      };
    case 'offsets':
      return {
        name: 'offsets',
        type: _l7Core.AttributeType.Attribute,
        descriptor: {
          name: 'a_Offsets',
          buffer: {
            // give the WebGL driver a hint that this buffer may change
            usage: _l7Core.gl.STATIC_DRAW,
            data: [],
            type: _l7Core.gl.FLOAT
          },
          size: 2,
          update: function update(feature) {
            var epo = feature.offsets;
            return epo;
          }
        }
      };
    default:
      return undefined;
  }
}
