// @ant-design/pro-card/es/components/Loading/index.js
import "antd/es/col/style";
import _Col from "antd/es/col";
import "antd/es/row/style";
import _Row from "antd/es/row";
import { jsx as _jsx, jsxs as _jsxs } from "react/jsx-runtime";
import './index.less';

var Loading = function Loading(props) {
  var style = props.style,
      prefix = props.prefix;
  return _jsxs("div", {
    className: "".concat(prefix, "-loading-content"),
    style: style,
    children: [_jsx(_Row, {
      gutter: 8,
      children: _jsx(_Col, {
        span: 22,
        children: _jsx("div", {
          className: "".concat(prefix, "-loading-block")
        })
      })
    }), _jsxs(_Row, {
      gutter: 8,
      children: [_jsx(_Col, {
        span: 8,
        children: _jsx("div", {
          className: "".concat(prefix, "-loading-block")
        })
      }), _jsx(_Col, {
        span: 15,
        children: _jsx("div", {
          className: "".concat(prefix, "-loading-block")
        })
      })]
    }), _jsxs(_Row, {
      gutter: 8,
      children: [_jsx(_Col, {
        span: 6,
        children: _jsx("div", {
          className: "".concat(prefix, "-loading-block")
        })
      }), _jsx(_Col, {
        span: 18,
        children: _jsx("div", {
          className: "".concat(prefix, "-loading-block")
        })
      })]
    }), _jsxs(_Row, {
      gutter: 8,
      children: [_jsx(_Col, {
        span: 13,
        children: _jsx("div", {
          className: "".concat(prefix, "-loading-block")
        })
      }), _jsx(_Col, {
        span: 9,
        children: _jsx("div", {
          className: "".concat(prefix, "-loading-block")
        })
      })]
    }), _jsxs(_Row, {
      gutter: 8,
      children: [_jsx(_Col, {
        span: 4,
        children: _jsx("div", {
          className: "".concat(prefix, "-loading-block")
        })
      }), _jsx(_Col, {
        span: 3,
        children: _jsx("div", {
          className: "".concat(prefix, "-loading-block")
        })
      }), _jsx(_Col, {
        span: 16,
        children: _jsx("div", {
          className: "".concat(prefix, "-loading-block")
        })
      })]
    })]
  });
};

export default Loading;
