// @ant-design/pro-layout/lib/components/PageLoading/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

var _objectSpread2 = _interopRequireDefault(require("@babel/runtime/helpers/objectSpread2"));

require("antd/lib/spin/style");

var _spin = _interopRequireDefault(require("antd/lib/spin"));

var _objectWithoutProperties2 = _interopRequireDefault(require("@babel/runtime/helpers/objectWithoutProperties"));

var _jsxRuntime = require("react/jsx-runtime");

var _excluded = ["isLoading", "pastDelay", "timedOut", "error", "retry"];

var PageLoading = function PageLoading(_ref) {
  var isLoading = _ref.isLoading,
      pastDelay = _ref.pastDelay,
      timedOut = _ref.timedOut,
      error = _ref.error,
      retry = _ref.retry,
      reset = (0, _objectWithoutProperties2.default)(_ref, _excluded);
  return (0, _jsxRuntime.jsx)("div", {
    style: {
      paddingTop: 100,
      textAlign: 'center'
    },
    children: (0, _jsxRuntime.jsx)(_spin.default, (0, _objectSpread2.default)({
      size: "large"
    }, reset))
  });
};

var _default = PageLoading;
exports.default = _default;
