// @ant-design/pro-card/lib/components/Loading/index.js
"use strict";

var _interopRequireDefault = require("@babel/runtime/helpers/interopRequireDefault").default;

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;

require("antd/lib/col/style");

var _col = _interopRequireDefault(require("antd/lib/col"));

require("antd/lib/row/style");

var _row = _interopRequireDefault(require("antd/lib/row"));

var _jsxRuntime = require("react/jsx-runtime");

require("./index.less");

var Loading = function Loading(props) {
  var style = props.style,
      prefix = props.prefix;
  return (0, _jsxRuntime.jsxs)("div", {
    className: "".concat(prefix, "-loading-content"),
    style: style,
    children: [(0, _jsxRuntime.jsx)(_row.default, {
      gutter: 8,
      children: (0, _jsxRuntime.jsx)(_col.default, {
        span: 22,
        children: (0, _jsxRuntime.jsx)("div", {
          className: "".concat(prefix, "-loading-block")
        })
      })
    }), (0, _jsxRuntime.jsxs)(_row.default, {
      gutter: 8,
      children: [(0, _jsxRuntime.jsx)(_col.default, {
        span: 8,
        children: (0, _jsxRuntime.jsx)("div", {
          className: "".concat(prefix, "-loading-block")
        })
      }), (0, _jsxRuntime.jsx)(_col.default, {
        span: 15,
        children: (0, _jsxRuntime.jsx)("div", {
          className: "".concat(prefix, "-loading-block")
        })
      })]
    }), (0, _jsxRuntime.jsxs)(_row.default, {
      gutter: 8,
      children: [(0, _jsxRuntime.jsx)(_col.default, {
        span: 6,
        children: (0, _jsxRuntime.jsx)("div", {
          className: "".concat(prefix, "-loading-block")
        })
      }), (0, _jsxRuntime.jsx)(_col.default, {
        span: 18,
        children: (0, _jsxRuntime.jsx)("div", {
          className: "".concat(prefix, "-loading-block")
        })
      })]
    }), (0, _jsxRuntime.jsxs)(_row.default, {
      gutter: 8,
      children: [(0, _jsxRuntime.jsx)(_col.default, {
        span: 13,
        children: (0, _jsxRuntime.jsx)("div", {
          className: "".concat(prefix, "-loading-block")
        })
      }), (0, _jsxRuntime.jsx)(_col.default, {
        span: 9,
        children: (0, _jsxRuntime.jsx)("div", {
          className: "".concat(prefix, "-loading-block")
        })
      })]
    }), (0, _jsxRuntime.jsxs)(_row.default, {
      gutter: 8,
      children: [(0, _jsxRuntime.jsx)(_col.default, {
        span: 4,
        children: (0, _jsxRuntime.jsx)("div", {
          className: "".concat(prefix, "-loading-block")
        })
      }), (0, _jsxRuntime.jsx)(_col.default, {
        span: 3,
        children: (0, _jsxRuntime.jsx)("div", {
          className: "".concat(prefix, "-loading-block")
        })
      }), (0, _jsxRuntime.jsx)(_col.default, {
        span: 16,
        children: (0, _jsxRuntime.jsx)("div", {
          className: "".concat(prefix, "-loading-block")
        })
      })]
    })]
  });
};

var _default = Loading;
exports.default = _default;
